lib/mempcpy.h

```diff
@@ -0,0 +1,32 @@
+/*
+ * SPDX-FileCopyrightText:  2023, Alejandro Colomar <alx@kernel.org>
+ *
+ * SPDX-License-Identifier:  BSD-3-Clause
+ */
+
+
+#ifndef SHADOW_INCLUDE_LIB_MEMPCPY_H_
+#define SHADOW_INCLUDE_LIB_MEMPCPY_H_
+
+
+#include <config.h>
+
+#include <stddef.h>
+#include <string.h>
+
+
+#if !defined(HAVE_MEMPCPY)
+inline void *mempcpy(void *restrict dst, const void *restrict src, size_t n);
+#endif
+
+
+#if !defined(HAVE_MEMPCPY)
+inline void *
+mempcpy(void *restrict dst, const void *restrict src, size_t n)
+{
+	return memcpy(dst, src, n) + n;
```

Comment:
For an `extern` function (most functions), yes, the implementation goes in a .c file.

However, `inline` functions have several benefits:

-  Performance (but for shadow-utils we don't really care about it).
-  The compiler (and static analyzers) can see what the function does,
   which allows more and better warnings.
-  It tells readers that the function is just a thin wrapper around another function,
   or a very small function otherwise (like the bit manipulation functions).

Since few projects use this modern C feature, I guess it feels a bit weird,
and many otherwise-very good programmers completely ignore C99 inline.
Which makes it also a good reason to use it, even if just to be an
example of good practice regarding using `inline`.

Reply:
I prefer `inline` over macros in most cases, since it offers more safety,
and better performance, but you can think of it as macros on steroids.

I restrict the use of macros for when you can't do the same thing with
functions (that is, when we need to use `sizeof()` of an argument, and
similar stuff.

Reply:
I had to refresh my memories of `inline` functions. Thanks for the explanation ;)

This [link](https://www.greenend.org.uk/rjk/tech/inline.html) can be used as a broader explanation.

Reply:
Indeed.  That link is the best resource for learning `inline` (actually, I learnt it from there).  Very recommended reading for every C programmer.

:)